File: api/model/cargos.py
"""api/model/cargos.py

Espelho Python da tabela canônica de cargos — `lib/config/cargos.ts`.

## Por que existe um espelho, e não um import

O orchestrator roda em Python (runtime Vercel `python3.14`, ver
`api/model/project.py`) e o catálogo de cargos é TypeScript. Não há
mecanismo de import entre os dois: o processo Python não carrega módulo TS,
e serializar a tabela em build time acrescentaria um passo de geração ao
`pnpm build` que nada mais no repositório usa.

O que este módulo NÃO é: uma segunda fonte de verdade. `lib/config/cargos.ts`
continua sendo a única — ele nasceu justamente para acabar com as quatro
cópias desalinhadas do mesmo conhecimento (ver o cabeçalho de lá). O que
protege a sincronia é um teste, não a boa vontade:
`tests/unit/model/test_cargos_sync.py` lê o `.ts` e falha se qualquer campo
aqui divergir do de lá. Um cargo novo, ou uma mudança de `vagas_por_uf`,
quebra o teste antes de chegar ao ar.

## O que o modelo lê daqui

Dois campos, e só:

  - `vagas_por_uf` — quantos eleitos a UF entrega. É o `vagas` de
    `p_vitoria.p_eleito` (RF-103): 1 para Presidente/Governador, **2** para
    Senador em 2026 (renovação de 2/3 → 54 vagas), `None` para Deputado
    Federal, que é proporcional e tem bancada variável por UF.
  - `granularidade` — `"zona"` ou `"uf"` (ADR-0026 item 1). Vai para
    `metodo.granularidade` no payload (RF-102): é a diferença que a tela de
    Senador precisa declarar ao leitor, porque a projeção dele NÃO é zona a
    zona como a de Presidente e Governador.

Constituição § 9: nenhum I/O, tabela estática e funções puras — igual ao
módulo TS que este espelha.
"""

from __future__ import annotations

from typing import Literal, TypedDict

Granularidade = Literal["uf", "zona"]


class CargoInfo(TypedDict):
    """Subconjunto de `CargoInfo` (TS) que o modelo consome.

    `token`, `slug` e `label` entram porque o teste de sincronia os confere —
    é barato manter a linha inteira alinhada e caro descobrir depois que só
    metade estava.
    """

    cd: int
    token: str
    slug: str
    label: str
    vagas_por_uf: int | None
    tem_segundo_turno: bool
    tem_arquivo_br: bool
    proporcional: bool
    granularidade: Granularidade


CARGOS: tuple[CargoInfo, ...] = (
    {
        "cd": 1,
        "token": "pres",
        "slug": "presidente",
        "label": "Presidente",
        "vagas_por_uf": 1,
        "tem_segundo_turno": True,
        "tem_arquivo_br": True,
        "proporcional": False,
        "granularidade": "zona",
    },
    {
        "cd": 3,
        "token": "gov",
        "slug": "governador",
        "label": "Governador",
        "vagas_por_uf": 1,
        "tem_segundo_turno": True,
        "tem_arquivo_br": False,
        "proporcional": False,
        "granularidade": "zona",
    },
    {
        "cd": 5,
        "token": "sen",
        "slug": "senador",
        "label": "Senador",
        "vagas_por_uf": 2,
        "tem_segundo_turno": False,
        "tem_arquivo_br": False,
        "proporcional": False,
        "granularidade": "zona",
    },
    {
        "cd": 6,
        "token": "dep",
        "slug": "deputado-federal",
        "label": "Deputado Federal",
        "vagas_por_uf": None,
        "tem_segundo_turno": False,
        "tem_arquivo_br": False,
        "proporcional": True,
        "granularidade": "zona",
    },
)

_POR_CD: dict[int, CargoInfo] = {c["cd"]: c for c in CARGOS}


# ---------------------------------------------------------------------------
# Fatos da eleição de 2026 — não são configuração, e não se derivam do ciclo
# ---------------------------------------------------------------------------

#: Cadeiras da casa legislativa inteira, por cargo. O Senado tem 81 (3 por UF
#: × 27); em 2026 renova 2/3 delas. A Câmara tem **513**, e em 2026 as 513
#: estão em disputa (ver `VAGAS_EM_DISPUTA_2026`). Presidente e Governador não
#: têm "casa" — ficam fora do dicionário.
#:
#: **513, e não 531 — desfecho conhecido desde julho/2025.** O PLP 177/2023
#: elevaria a Câmara a 531 pela redistribuição do Censo 2022; foi aprovado pela
#: Câmara e pelo Senado em junho/2025, **vetado integralmente pela Presidência
#: da República em julho/2025**, e o **STF decidiu manter a distribuição atual
#: de 513** para este pleito, que é regido pela Resolução TSE 23.751/2026
#: (`docs/reference/regulatory.md`). Até 2026-09-19 vários documentos deste
#: repositório ainda descreviam o desfecho como "não confirmado" e derivavam
#: o total do runtime por causa disso; a premissa morreu, a decisão de ler o
#: número **por UF** do TSE não (RF-124 — ver `VAGAS_EM_DISPUTA_2026`).
TOTAL_CADEIRAS: dict[int, int] = {5: 81, 6: 513}

#: Cadeiras que a eleição de 2026 renova, por cargo. **54** para o Senado —
#: 2 por UF × 27 — e **513** para a Câmara, que renova **integralmente**. Essa
#: é a diferença entre as duas casas, e é por isso que o cargo 6 repete aqui o
#: número de `TOTAL_CADEIRAS` e o cargo 5 não: o Senado renova 2/3 de 81, a
#: Câmara renova 513 de 513.
#:
#: É o denominador que a tela nacional de Senador exibe (RF-107), e ele não
#: pode ser derivado das UFs que já apuraram: às 18h, com 4 estados apurados, a
#: derivação diria "8 vagas em disputa", o que é falso. O número é fixo desde
#: antes da urna abrir.
#:
#: O mesmo argumento vale, palavra por palavra, para a Câmara — e lá ele não é
#: hipotético: até 2026-09-19 `deputado_payload._bancada_nacional` somava
#: `lugares_a_preencher` só das UFs presentes, e com três estados pequenos
#: apurados a tela escrevia "26 cadeiras em disputa". RF-124 continua regendo o
#: número **por UF**, que segue saindo do dado publicado pelo TSE; o total
#: nacional é fato fixo, **conferido** contra a soma quando as 27 UFs tiverem
#: publicado o seu `carg[].nv` (`deputado_payload.conferir_total_de_cadeiras`).
VAGAS_EM_DISPUTA_2026: dict[int, int] = {5: 54, 6: 513}


# ---------------------------------------------------------------------------
# Cadência de ingestão por cargo — base do limiar de "dado parado" (ADR-0038 D3)
# ---------------------------------------------------------------------------

#: Cadência declarada da corrida proporcional, em minutos (RF-128, ADR-0026
#: item 5). Entra no payload para que a tela não a escreva à mão — foi assim
#: que quatro frases do Senador viraram falsas em 11/09.
#:
#: 30, não 15, desde 2026-09-13: o cargo 6 saiu de granularidade UF (um cron
#: `*/15`) para ZONA fatiada em 6 (`sliceTargets`, `lib/tse/targets.ts`) — os
#: crons de `vercel.ts` disparam uma fatia a cada 5 min, e a volta completa
#: das 6 fatias (garantia de que toda UF foi revisitada) leva 30 min, não 15.
#:
#: Mudou de casa em 2026-09-13 (ADR-0038 D3): nasceu em `api/model/project.py`
#: e continua reexportado de lá (`from api.model.cargos import
#: ATUALIZACAO_MIN_DEPUTADO`), então nenhum uso anterior quebra. Veio para cá
#: porque o limiar de alarme precisa **derivar** dela — repetir "30" em
#: `api/model/dado_ts.py` seria o segundo número a manter à mão, exatamente o
#: que o ADR proíbe — e `dado_ts.py` não pode importar de `project.py`, que é
#: quem importa `dado_ts`.
ATUALIZACAO_MIN_DEPUTADO = 30

#: Cadência de ingestão de cada cargo, em **segundos**. É o intervalo entre
#: duas leituras COMPLETAS do universo de alvos daquele cargo — não o intervalo
#: entre dois disparos de cron.
#:
#: A distinção só importa no cargo 6, e importa muito: `vercel.ts:192-240` tem
#: seis crons de 5 em 5 minutos, um por fatia (`/api/ingest/deputado-federal/
#: 1..6`). Derivar "300s" desse `*/5` diria que toda zona é revisitada a cada
#: 5 min, quando na verdade cada uma é revisitada uma vez por volta completa —
#: 30 min (ADR-0036). Um limiar de alarme calibrado nos 5 min gritaria em todo
#: ciclo saudável de Deputado.
#:
#: Cargo desconhecido fica FORA do dicionário de propósito: sem cadência
#: declarada não há limiar honesto a aplicar, e `cadencia_segundos` devolve
#: `None` em vez de um default que alarmaria no ritmo errado.
CADENCIA_SEGUNDOS: dict[int, int] = {
    1: 60,  # Presidente — ADR-0011 (cadência de 60s)
    3: 60,  # Governador — ADR-0011, mesmo cron
    5: 300,  # Senador — ADR-0026 nota (b): 5 min, mantidos na volta para zona
    6: ATUALIZACAO_MIN_DEPUTADO * 60,  # Deputado — ADR-0036: 6 fatias × 5 min
}


def total_cadeiras(cd: int) -> int | None:
    """Tamanho da casa legislativa do cargo, ou `None` quando não se aplica."""
    return TOTAL_CADEIRAS.get(int(cd))


def vagas_em_disputa(cd: int) -> int | None:
    """Cadeiras renovadas em 2026, ou `None` quando não se aplica."""
    return VAGAS_EM_DISPUTA_2026.get(int(cd))


def cadencia_segundos(cd: int) -> int | None:
    """Intervalo entre duas varreduras COMPLETAS do cargo, em segundos.

    `None` para cargo sem cadência declarada — o chamador não tem limiar a
    aplicar e não deve inventar um (ver `CADENCIA_SEGUNDOS`).
    """
    return CADENCIA_SEGUNDOS.get(int(cd))


def cargo_info(cd: int) -> CargoInfo | None:
    """Metadados do cargo, ou `None` quando o código não é coberto.

    `None` em vez de exceção: o orchestrator aceita `cargo` de 1 a 99 no corpo
    do trigger (`ProjectRequest`) e precisa degradar — um cargo desconhecido
    cai no comportamento default (1 vaga, granularidade de zona), nunca
    derruba o ciclo.
    """
    return _POR_CD.get(int(cd))


def vagas_por_uf(cd: int, default: int = 1) -> int:
    """Vagas em disputa por UF. `default` cobre cargo desconhecido e o
    proporcional (Deputado Federal, `vagas_por_uf = None`), cuja bancada vem
    de tabela própria e não desta."""
    info = cargo_info(cd)
    if info is None:
        return default
    vagas = info["vagas_por_uf"]
    return default if vagas is None else vagas


def granularidade(cd: int, default: Granularidade = "zona") -> Granularidade:
    """Granularidade de ingestão do cargo (ADR-0026 item 1)."""
    info = cargo_info(cd)
    return info["granularidade"] if info is not None else default

File: lib/config/cargos.ts
/**
 * lib/config/cargos.ts
 *
 * **Mapa único dos cargos cobertos pelo SalaCofre.** Antes de 2026-09-11 esse
 * conhecimento estava espalhado por quatro lugares independentes, todos travados
 * em `1 | 3` e sem relação entre si:
 *
 *   - `lib/tse/targets.ts` — `Target.cargo`, `VALID_CARGOS`, `getActiveCargos`,
 *     `formatCargo`, `parseWhitelist` (código numérico do TSE);
 *   - `lib/edge-config/types.ts` — `Cargo = 1 | 3` (código numérico, no payload);
 *   - `lib/config/calendar.ts` — `Cargo = "pres" | "gov"` (token de chave, ADR-0012);
 *   - `app/api/ingest/[cargo]/route.ts` — `parseCargoSegment`, o único lugar que
 *     mapeava slug de URL para código.
 *
 * Estender a cobertura para Senador e Deputado Federal (ADR-0026) exigiria editar
 * os quatro em sincronia, sem nada que forçasse a sincronia. Este módulo é a
 * fonte única; os outros passam a derivar dele.
 *
 * ## O que NÃO muda
 *
 * Os **dois tipos `Cargo`** continuam separados de propósito (ADR-0026 item 2,
 * ADR-0028 item 2): o numérico (`CargoTse`) espelha o TSE e viaja no payload; o
 * de token (`lib/config/calendar.ts`) nomeia chaves do Global Config e caminhos
 * do Blob. Este módulo não os funde — dá o conversor explícito (`cargoToken`,
 * `cargoFromToken`) que antes não existia.
 *
 * Constituição § 9: nenhum I/O aqui. Tabela estática, funções puras.
 */

import type { Cargo as CargoToken } from "@/lib/config/calendar";

/**
 * Código numérico do cargo, como o TSE publica (campo `cd` em `carg[]` do EA20,
 * e o `-c<cargo4>` no nome do arquivo).
 *
 * 1 = Presidente · 3 = Governador · 5 = Senador · 6 = Deputado Federal.
 *
 * Os códigos 2 (Vice-Presidente), 4 (Vice-Governador) e 7/8 (Deputado Estadual /
 * Distrital) existem no TSE e **não** são cobertos: vice não tem votação própria,
 * e as assembleias estaduais estão fora do escopo do produto.
 */
export type CargoTse = 1 | 3 | 5 | 6;

/**
 * Qual das DUAS eleições do pleito 2026 cobre este cargo (parâmetros
 * publicados pelo TSE em 17/09/2026, véspera do simulado): a Eleição
 * Federal (código `21270`) elege só o Presidente; a Eleição Estadual
 * (código `21272`) elege Governador, Senador e Deputado Federal. Os dois
 * códigos vivem no MESMO `codEleicao` (`ele<AAAA>/<dígitos>`, ver
 * `lib/tse/targets.ts::getCodEleicao`) — não são turnos nem ambientes, são
 * duas árvores de URL paralelas no CDN (`.../ele2026/21270/dados/...` vs
 * `.../ele2026/21272/dados/...`), inclusive para o EA14 de acompanhamento
 * (um `br-e021270-ab.json`, outro `br-e021272-ab.json`).
 */
export type Eleicao = "federal" | "estadual";

/** Metadados de um cargo coberto. */
export interface CargoInfo {
  /** Código do TSE. */
  readonly cd: CargoTse;
  /**
   * Eleição (no sentido do pleito 2026, ver `Eleicao`) a que este cargo
   * pertence — determina qual código de eleição resolver
   * (`getCodEleicaoDoCargo`, `lib/tse/targets.ts`). Presidente é o único
   * cargo `"federal"`; os outros três são `"estadual"`. Campo obrigatório
   * de propósito: não há valor "neutro" que sirva de default seguro — um
   * cargo estadual lido como federal (ou vice-versa) busca o EA20 do CÓDIGO
   * DE ELEIÇÃO ERRADO, um 404 sistemático que não é óbvio de diagnosticar.
   */
  readonly eleicao: Eleicao;
  /** Token de namespacing de chave (Global Config / Blob) — ADR-0012. */
  readonly token: CargoToken;
  /** Segmento de rota aceito por `/api/ingest/[cargo]` e pelas páginas. */
  readonly slug: string;
  /** Rótulo para leitor humano. */
  readonly label: string;
  /**
   * Vagas em disputa por UF. Presidente e Governador elegem 1 por abrangência;
   * o Senado renova **2/3** em 2026, o que são **2 vagas por UF** (54 no total)
   * — não 1, apesar de o kit de UI rotular "1 vaga"
   * (ver `docs/architecture/adrs/0029-home-mobile-first-mapa-primeiro-fiel-ao-kit.md`).
   * Deputado Federal é proporcional: o nº de cadeiras varia por UF (8 a 70), então
   * fica `null` aqui e vem da tabela de bancadas.
   */
  readonly vagasPorUf: number | null;
  /** `true` quando a disputa admite 2º turno. Senador e Deputado são turno único. */
  readonly temSegundoTurno: boolean;
  /**
   * `true` só para Presidente: é o único cargo com arquivo agregado nacional
   * (`br-c0001-...`) no CDN do TSE. Os demais só têm UF/município/zona, e o
   * agregado nacional deles é soma nossa.
   */
  readonly temArquivoBr: boolean;
  /** `true` quando o cargo é proporcional (votos viram cadeiras via quociente). */
  readonly proporcional: boolean;
  /**
   * Granularidade de ingestão **padrão** deste cargo (ADR-0026 item 1).
   *
   * `"zona"` (~6.110 pares por cargo) é o que o estimador precisa para a
   * regra de três (ADR-0021, RF-011/012) — é o modo de Presidente e
   * Governador, e desde 2026-09-11 também de Senador e Deputado Federal.
   *
   * **Senador saiu de `"uf"` para `"zona"` em 2026-09-11**, emendando o
   * ADR-0026 item 1. Motivo medido: com um único boletim por estado, o bootstrap
   * do estimador tem uma só unidade de reamostragem — as 1.000 réplicas saem
   * idênticas, o IC95 fecha num ponto e `p_eleito` degenera para 0% ou 100%.
   * Verificado à parte: 1 observação produz **1** réplica distinta; 3 produzem
   * 10. Exibir aquilo como probabilidade afirmaria certeza que o modelo não tem
   * (constituição § 6), e a alternativa era publicar o cargo sem chance de
   * eleição. Decisão do usuário, com o custo aceito de baixar os três cargos
   * pesados de 35 para 25 rps.
   *
   * **Deputado Federal saiu de `"uf"` para `"zona"` em 2026-09-13**, mesmo
   * diagnóstico (ADR-0026, nota "2026-09-11 (b)"): um único arquivo por UF só
   * dá ao bootstrap do RF-127 uma unidade de reamostragem, e o IC95 degenera
   * do mesmo jeito. A diferença para o Senador é o volume: ~6.110 alvos a
   * `rpsMax=5` (ver abaixo) levariam ~1.222 s numa invocação só, muito acima
   * do `maxDuration` de 300 s — por isso a varredura é dividida em 6 fatias
   * (`/api/ingest/deputado-federal/<1..6>`, `sliceTargets` em
   * `lib/tse/targets.ts`), cada uma cobrindo ~1/6 do fan-out (~1.019 alvos,
   * ~204 s), disparadas a cada 5 min — a volta completa das 6 fatias leva
   * 30 min. Interruptor de emergência específico deste cargo, sem deploy:
   * `TSE_DEPUTADO_GRANULARIDADE=uf` (`lib/tse/targets.ts::getGranularidade`,
   * documentado em `docs/operations/runbook.md` § Variáveis de ambiente).
   *
   * `TSE_GRANULARIDADE` no ambiente sobrepõe isto para TODOS os cargos —
   * é escotilha de diagnóstico, não configuração de produção.
   */
  readonly granularidade: "uf" | "zona";
  /**
   * Teto de requisições por segundo **deste cargo**, quando `TSE_MAX_RPS` não
   * está definida no ambiente (constituição § 1).
   *
   * Por que é por cargo, e não um número só: os quatro crons podem disparar no
   * MESMO minuto — a cadência de 5 minutos (Senador, e cada fatia de Deputado
   * Federal desde o ADR-0036) cai sobre a de 1 minuto de Presidente e
   * Governador **a cada múltiplo de 5**, não só nos minutos 0/15/30/45 como
   * quando Deputado era um cron único de 15 em 15 min. Cada invocação tem seu
   * (o rate limiter é singleton **de processo**, e o Fluid Compute isola
   * instâncias), então o que o TSE vê no IP é a **soma**.
   *
   * As 6 fatias do cargo 6 são **intercaladas** em `vercel.ts` (fatia 1 nos
   * minutos 0 e 30, fatia 2 em 5 e 35, ..., fatia 6 em 25 e 55), de modo que
   * no máximo UMA delas está no ar por vez. É isso que mantém a contribuição
   * do Deputado em 5 rps e o agregado em 80 — seis fatias simultâneas dariam
   * 30 rps só dele, e 105 no total.
   *
   * Medido em 2026-09-11, com todos em 40: pico de **160 rps** com quatro
   * simultâneos e **120** com três — acima do teto documentado de 100, que
   * bloqueia o IP por 10 minutos. O default de 40 tinha sido calibrado para
   * DOIS cargos (2 x 40 = 80) e não sobreviveu à entrada de Senador e Deputado.
   *
   * Calibragem atual — pior caso agregado **80 rps**, 20% abaixo do teto:
   *
   *   | cargo      | alvos | rps | duração do ciclo                        |
   *   |------------|-------|-----|------------------------------------------|
   *   | Presidente | 6.110 |  25 | ~244 s                                   |
   *   | Governador | 6.110 |  25 | ~244 s                                   |
   *   | Senador    | 6.110 |  25 | ~244 s                                   |
   *   | Deputado   | 6.110 |   5 | ~1.222 s inteiro; ~204 s POR FATIA (÷6)  |
   *
   * Os pesados caíram de 35 para 25 rps em 2026-09-11, quando Senador passou a
   * ser ingerido por zona (decisão do usuário — ver `granularidade`): três
   * cargos pesados a 35 dariam 110 rps agregados, acima do teto. A 25, o ciclo
   * mais longo vai a ~244 s, dentro do `maxDuration` de 300 s mas com menos
   * folga que antes — é o custo explícito da decisão, e o que torna a medição
   * de `duration_ms` no simulado 1 obrigatória, não opcional.
   *
   * **Deputado Federal NÃO mudou de rps em 2026-09-13**, quando saiu de UF
   * para zona (ver `granularidade`) — continua em 5, deliberadamente: o
   * orçamento agregado do IP (80 rps) já está comprometido pelos três
   * pesados, e Deputado é o cargo com menor urgência editorial dos quatro. O
   * que mudou foi dividir os ~6.110 alvos em 6 fatias por invocação
   * (`sliceTargets`, `lib/tse/targets.ts`) em vez de pedir todos numa
   * invocação só: a 5 rps, uma fatia de ~1.019 alvos leva ~204 s, dentro do
   * `maxDuration` de 300 s — o total sem fatiar (~1.222 s) não caberia.
   */
  readonly rpsMax: number;
}

/**
 * Tabela canônica. A ordem é a de exibição nas abas
 * (`components/layout/CargoTabs.tsx`), não a numérica do TSE.
 */
export const CARGOS: readonly CargoInfo[] = [
  {
    cd: 1,
    eleicao: "federal",
    token: "pres",
    slug: "presidente",
    label: "Presidente",
    vagasPorUf: 1,
    temSegundoTurno: true,
    temArquivoBr: true,
    proporcional: false,
    granularidade: "zona",
    rpsMax: 25,
  },
  {
    cd: 3,
    eleicao: "estadual",
    token: "gov",
    slug: "governador",
    label: "Governador",
    vagasPorUf: 1,
    temSegundoTurno: true,
    temArquivoBr: false,
    proporcional: false,
    granularidade: "zona",
    rpsMax: 25,
  },
  {
    cd: 5,
    eleicao: "estadual",
    token: "sen",
    slug: "senador",
    label: "Senador",
    vagasPorUf: 2,
    temSegundoTurno: false,
    temArquivoBr: false,
    proporcional: false,
    granularidade: "zona",
    rpsMax: 25,
  },
  {
    cd: 6,
    eleicao: "estadual",
    token: "dep",
    slug: "deputado-federal",
    label: "Deputado Federal",
    vagasPorUf: null,
    temSegundoTurno: false,
    temArquivoBr: false,
    proporcional: true,
    granularidade: "zona",
    rpsMax: 5,
  },
] as const;

/** Todos os códigos cobertos, na ordem da tabela. */
export const CARGOS_TSE: readonly CargoTse[] = CARGOS.map((c) => c.cd);

const POR_CD = new Map<number, CargoInfo>(CARGOS.map((c) => [c.cd, c]));
const POR_TOKEN = new Map<string, CargoInfo>(CARGOS.map((c) => [c.token, c]));
const POR_SLUG = new Map<string, CargoInfo>(CARGOS.map((c) => [c.slug, c]));

/** Type guard — `true` se o número é um cargo coberto. */
export function isCargoTse(n: number): n is CargoTse {
  return POR_CD.has(n);
}

/** Metadados do cargo, ou `undefined` se não for coberto. */
export function cargoInfo(cd: CargoTse): CargoInfo {
  const info = POR_CD.get(cd);
  // Inalcançável pelo tipo; a guarda existe para o caso de um `as CargoTse`
  // indevido em código de borda.
  if (!info) throw new Error(`[cargos] código de cargo não coberto: ${cd}`);
  return info;
}

/** Código numérico do TSE → token de chave (`1` → `"pres"`). */
export function cargoToken(cd: CargoTse): CargoToken {
  return cargoInfo(cd).token;
}

/** Token de chave → código numérico (`"pres"` → `1`). */
export function cargoFromToken(token: CargoToken): CargoTse {
  const info = POR_TOKEN.get(token);
  if (!info) throw new Error(`[cargos] token de cargo desconhecido: ${token}`);
  return info.cd;
}

/**
 * Eleição (federal/estadual, ver `Eleicao`) a que este cargo pertence —
 * lê direto da tabela canônica, sem ternário nem `??`: Presidente (1) é
 * `"federal"`, os outros três são `"estadual"`. Usado por
 * `lib/tse/targets.ts::getCodEleicaoDoCargo` para resolver qual dos dois
 * códigos de eleição do pleito 2026 (`21270` federal, `21272` estadual)
 * corresponde a um cargo.
 */
export function eleicaoDoCargo(cd: CargoTse): Eleicao {
  return cargoInfo(cd).eleicao;
}

/**
 * Resolve o segmento de rota de `/api/ingest/[cargo]` e das páginas de cargo.
 *
 * Aceita o **código** (`"1"`, `"5"`) ou o **slug** (`"presidente"`,
 * `"deputado-federal"`), case-insensitive. Devolve `null` para qualquer outra
 * coisa — o caller responde 400, sem revelar quais valores existem.
 */
export function parseCargoSegment(raw: string): CargoTse | null {
  const s = raw.trim().toLowerCase();
  if (!s) return null;
  const porSlug = POR_SLUG.get(s);
  if (porSlug) return porSlug.cd;
  const n = Number(s);
  return Number.isInteger(n) && isCargoTse(n) ? n : null;
}

/**
 * Pior caso agregado de requisições por segundo contra o IP do TSE: todos os
 * cargos cobertos disparando ao mesmo tempo.
 *
 * Não é hipótese — os crons de `vercel.ts` coincidem **a cada 5 minutos**,
 * porque a cadência de 5 min do Senador e das fatias de Deputado Federal cai
 * sobre a de 1 min dos majoritários. O teto documentado do TSE é 100 rps por
 * IP, com bloqueio de
 * 10 minutos, e a constituição § 1 exige margem **bem abaixo** disso, não
 * "exatamente no limite".
 */
export function piorCasoAgregadoRps(): number {
  return CARGOS.reduce((acc, c) => acc + c.rpsMax, 0);
}

/**
 * Teto de rps de um ciclo que cobre VÁRIOS cargos no mesmo processo.
 *
 * O ciclo genérico (`/api/ingest`, sem segmento) percorre os cargos
 * sequencialmente dentro de uma invocação, com **um** bucket. Sua contribuição
 * ao IP é a de um processo só, então ele pode usar o maior teto entre os cargos
 * que cobre — não a soma, e não o menor (que arrastaria o fan-out pesado a
 * 1.222 s, muito além do `maxDuration`).
 */
export function rpsMaxParaCargos(cargos: readonly CargoTse[]): number {
  if (cargos.length === 0) return Math.min(...CARGOS.map((c) => c.rpsMax));
  return Math.max(...cargos.map((c) => cargoInfo(c).rpsMax));
}
